libcxx/include/flat_set

```diff
@@ -0,0 +1,913 @@
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t __s, initializer_list<value_type> __il,
+           const type_identity_t<key_compare>& __comp, const _Alloc& __a)
+    : flat_set(__s, __il.begin(), __il.end(), __comp, __a) {}
+
+  template <class _Alloc>
+    requires uses_allocator_v<container_type, _Alloc>
+  flat_set(sorted_unique_t __s, initializer_list<value_type> __il, const _Alloc& __a)
+    : flat_set(__s, __il.begin(), __il.end(), __a) { }
+
+  flat_set& operator=(initializer_list<value_type> __il) {
+    clear();
+    insert(__il.begin(), __il.end());
+    return *this;
+  }
+
+  flat_set& operator=(const flat_set& __fs) {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    __compare_ = __fs.__compare_;
+    __c_ = __fs.__c_;
+    __guard.__complete();
+    return *this;
+  }
+
+  flat_set& operator=(flat_set&& __fs)
+    noexcept(is_nothrow_assignable_v<_Compare&, _Compare&> && is_nothrow_assignable_v<_KeyContainer, _KeyContainer&&>)
+  {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    __compare_ = __fs.__compare_;
+    __c_ = std::move(__fs.__c_);
+    __guard.__complete();
+    return *this;
+  }
+
+  // iterators
+
+  iterator begin() noexcept { return __c_.begin(); }
+  const_iterator begin() const noexcept { return __c_.begin(); }
+  iterator end() noexcept { return __c_.end(); }
+  const_iterator end() const noexcept { return __c_.end(); }
+
+  reverse_iterator rbegin() noexcept { return reverse_iterator(end()); }
+  const_reverse_iterator rbegin() const noexcept { return const_reverse_iterator(end()); }
+  reverse_iterator rend() noexcept { return reverse_iterator(begin()); }
+  const_reverse_iterator rend() const noexcept { return const_reverse_iterator(begin()); }
+
+  const_iterator cbegin() const noexcept { return begin(); }
+  const_iterator cend() const noexcept { return end(); }
+  const_reverse_iterator crbegin() const noexcept { return rbegin(); }
+  const_reverse_iterator crend() const noexcept { return rend(); }
+
+  // capacity
+
+  [[nodiscard]] bool empty() const noexcept { return __c_.empty(); }
+  size_type size() const noexcept { return __c_.size(); }
+  size_type max_size() const noexcept { return __c_.max_size(); }
+
+  // [flat.set.modifiers]
+
+  template <class... _Args>
+  pair<iterator, bool> emplace(_Args&&... __args) {
+    auto __key = key_type(std::forward<_Args>(__args)...);
+    auto __it = lower_bound(__key);
+    if (__it == end() || __compare_(__key, *__it)) {
+      auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+      __it = __c_.emplace(__it, std::move(__key));
+      __guard.__complete();
+      return {__it, true};
+    } else {
+      return {__it, false};
+    }
+  }
+
+  template <class... _Args>
+  iterator emplace_hint(const_iterator, _Args&&... __args) {
+    return emplace(std::forward<_Args>(__args)...).first;
```

Comment:
I think we can actually use the hint. If the user gives us an iterator to the lower bound, then we can avoid computing it ourselves and perform the insertion directly at that location. You might be able to implement `emplace` in terms of `emplace_hint` in that case.

Reply:
Done. I factored out an `__emplace_unchecked(pos, K2&& x)` that assumes the insertion _does_ happen, at the _exact_ location passed to it; all the insert/emplace sites use that now.

FWIW, I see that back in 2019 @tzlaine and I discussed this exact API (on the cpplang Slack) and Zach was at least briefly considering not to provide it at all. His reasoning was:
- `mySet.insert(x)` is O(lg n), but `mySet.insert(pos, x)` with the right hint is O(1).
- `myFlatSet.insert(x)` is O(n), but `myFlatSet.insert(pos, x)` with the right hint is... still O(n).
Does that change your calculus at all? (Of course at this point I've already done the implementation... but it's still a fair bit of code to maintain and test, and I admit I haven't written any good tests for all the `emplace_hint` corner cases yet. Ignoring the hint would be slightly less code and way fewer obscure branches to test.)